speaker/main.go

```diff
@@ -281,78 +289,109 @@ func (c *controller) SetBalancer(l log.Logger, name string, svc *v1.Service, eps
 	}
 
 	if deleteReason := handler.ShouldAnnounce(l, name, lbIPs, svc, eps); deleteReason != "" {
-		return c.deleteBalancer(l, name, deleteReason)
+		return c.deleteBalancerProtocol(l, protocol, name, deleteReason)
 	}
 
 	if err := handler.SetBalancer(l, name, lbIPs, pool); err != nil {
 		level.Error(l).Log("op", "setBalancer", "error", err, "msg", "failed to announce service")
 		return controllers.SyncStateError
 	}
 
-	if c.announced[name] == "" {
-		c.announced[name] = pool.Protocol
+	if !c.announced[protocol][name] {
+		c.announced[protocol][name] = true
 		c.svcIPs[name] = lbIPs
 	}
 
 	for _, ip := range lbIPs {
 		announcing.With(prometheus.Labels{
-			"protocol": string(pool.Protocol),
+			"protocol": string(protocol),
 			"service":  name,
 			"node":     c.myNode,
 			"ip":       ip.String(),
 		}).Set(1)
 	}
-	level.Info(l).Log("event", "serviceAnnounced", "msg", "service has IP, announcing")
-	c.client.Infof(svc, "nodeAssigned", "announcing from node %q", c.myNode)
-
+	level.Info(l).Log("event", "serviceAnnounced", "msg", "service has IP, announcing", "protocol", protocol)
+	c.client.Infof(svc, "nodeAssigned", "announcing from node %q with protocol %q", c.myNode, protocol)
 	return controllers.SyncStateSuccess
 }
 
 func (c *controller) deleteBalancer(l log.Logger, name, reason string) controllers.SyncState {
-	proto, ok := c.announced[name]
-	if !ok {
+	for _, protocol := range c.protocols {
+		if st := c.deleteBalancerProtocol(l, protocol, name, reason); st == controllers.SyncStateError {
+			return st
+		}
+	}
+	return controllers.SyncStateSuccess
+}
+
+func (c *controller) deleteBalancerProtocol(l log.Logger, protocol config.Proto, name, reason string) controllers.SyncState {
+	announced := c.announced[protocol][name]
+	if !announced {
 		return controllers.SyncStateSuccess
 	}
 
-	if err := c.protocols[proto].DeleteBalancer(l, name, reason); err != nil {
-		level.Error(l).Log("op", "deleteBalancer", "error", err, "msg", "failed to clear balancer state")
+	if err := c.protocolHandlers[protocol].DeleteBalancer(l, name, reason); err != nil {
+		level.Error(l).Log("op", "deleteBalancer", "error", err, "msg", "failed to clear balancer state", "protocol", protocol)
 		return controllers.SyncStateError
 	}
 
 	for _, ip := range c.svcIPs[name] {
 		ok := announcing.Delete(prometheus.Labels{
-			"protocol": string(proto),
+			"protocol": string(protocol),
 			"service":  name,
 			"node":     c.myNode,
 			"ip":       ip.String(),
 		})
 		if !ok {
-			level.Error(l).Log("op", "deleteBalancer", "error", "failed to delete service metric", "service", name, "ip", ip.String())
+			level.Error(l).Log("op", "deleteBalancer", "error", "failed to delete service metric", "service", name, "protocol", protocol, "ip", ip.String())
 		}
 	}
-	delete(c.announced, name)
-	delete(c.svcIPs, name)
+	delete(c.announced[protocol], name)
 
+	// we withdraw the service only if we are removing it from the last protocol
```

Comment:
can/do we test this somewhere?

Reply:
Added a unit test for checking this part of the code